### ISSUE TITLE
Encourage users to upload wheels for pure Python packages

### ISSUE THREAD

**Issue by rth** [pypa/packaging-problems#496]
I was wondering if there is any mechanism that would allow encouraging users to upload wheels for pure Python packages? Unless this should be a feature request for https://github.com/pypa/warehouse?

**Problem statement**

Currently some package maintainers only upload a .tar.gz to PyPi for pure Python packages, without the associated wheel. For instance, according to the analysis in https://gist.github.com/josephrocca/ca3f09c4db1df6288bbc64428e6bed77 around 1/4 of 100 most frequently downloaded PyPi packages only include the source distribution. Some maintainers might forget, other may not be aware of the motivation to do so.

This results in,
 - slightly slower installs since the wheel is then built locally (which is fast for pure Python packages, but still adds some overhead)
 - harder to analyze package artifacts (with a dynamic behavior in `setup.py`)
 - and in particular case of https://github.com/pyodide/pyodide inability to install the package from PyPi (https://github.com/pyodide/pyodide/issues/1501). We keep asking maintainers to add wheels but approach doesn't scale, and we were wondering if some centralized solution might exist that would benefit everyone, and slowly move the  ecosystem in that direction.

For instance `poetry publish` would already upload both the `.tar.gz` for pure Python packages.

**Possible implementation**

Given that artifacts can be uploaded with multiple calls, I'm not sure if it's something that would be technically possible even in case of interest in this feature.

One possibility could be to display a note/warning, if twine is used to upload the sdist and the corresponding wheel is not present for pure Python packages. Though that would likely produce false positives.

Another possibility could be to emphasize this a bit more in the documentation.

**Comment by di** [comment]
I think this is probably a duplicate of #25, which I don't think is exclusive to wheels with C extensions.

I think the "Create a generic wheel-building service to make releases faster and more robust" part of https://github.com/psf/fundable-packaging-improvements/blob/master/FUNDABLES.md is also relevant here.

**Comment by henryiii** [comment]
Build creates both wheel and SDist by default, so that helps here I think. And tools like cibuildwheel helps making binary wheels (ideally) on CI if it's not pure Python. As long as docs point to making and uploading wheels too, which I think ours do, I don't know if there's that much that can change.

**Comment by rth** [comment]
> I think the "Create a generic wheel-building service to make releases faster and more robust" part of https://github.com/psf/fundable-packaging-improvements/blob/master/FUNDABLES.md is also relevant here.

If PyPi could build the wheel automatically for pure Python packages, that would clearly be the ideal case. But I imagine that would take lots of effort and funding.;.

> #25, which I don't think is exclusive to wheels with C extensions.

It's not, but the subset of that issue discussed here, should also be less difficult to address from a technical/UX point of view (since it doesn't need to deal with the building C extensions discussion).

> Build creates both wheel and SDist by default, so that helps here I think.

What command do you mean by "build"? Yeah, if the default workflow was
```bash
twine build   # generates sdist + wheel for pure Python packages
twine upload  # without specifying the path, upload both
```
(or using some other tool) then most of this issue would be addressed IMO.

**Comment by henryiii** [comment]
[`pypa/build`](https://github.com/pypa/build), that is,

```python
pip install build
python -m build
```

or

```python
pipx run build
```

(Which works on GHA out of the box with no Python setup!)

See https://packaging.python.org/tutorials/packaging-projects/#generating-distribution-archives or https://scikit-hep.org/developer/gha_pure for examples.

**Comment by pfmoore** [comment]
> What command do you mean by "build"?

https://pypi.org/project/build/

That's the build tool recommended in [the tutorial](https://packaging.python.org/tutorials/packaging-projects/#generating-distribution-archives).

**Comment by rth** [comment]
That's great, thanks! I missed that development.

Just need support for,
```bash
twine upload   # without a path
```
then...

**Comment by rth** [comment]
To elaborate, I'm not sure this applies to other maintainers, but I would probably not run
```
twine upload dist/*
```
as recommended in the tutorial, but rather would upload them one by one separately to make sure no other unrelated packages get uploaded (for instance `python -m wheel` generates wheels for dependencies for packages with C extensions). Or at least carefully check that folder first. So some logic that does the correct version filtering by default without relying on bash globing might help. As `poetry publish` does I imagine.

**Comment by henryiii** [comment]
> If PyPi could build the wheel automatically for pure Python packages

This would be a mess; there [are pure Python packages](https://github.com/facelessuser/wcmatch) that depend on the Python version to generate a wheel, for example; so it it was done "automatically", then the wheel would only work on one version of Python. [Most packages](https://pythonwheels.com) do provide wheels; the ones that don't usually have a reason they are working on it. `black` didn't have a wheel for the previous version, but that's because they deleted the wheel since there was a possible problem with it (package and module both present). Missing a wheel when it's easy to add one _does_ happen, but it's not that bad - most tutorials make wheels too, and that's where it counts.

> twine upload without a path

Most tools (pip, build, gh-action-pypi-publish) assume `dist/*` here if no path is given, so seems to be a reasonable request (to be opened in twine). But `twine upload dist/*` isn't that hard to type either, and is in all the tutorials have it that way (I think).

**Comment by henryiii** [comment]
> for instance python -m wheel generates wheels for dependencies for packages

That's because this is being misused, you should use `build` instead, which does not generate extra packages, because it's not setting up a wheelhouse, but instead is building packages.

**Comment by henryiii** [comment]
> generates wheels for dependencies for packages with C extensions

Not exactly, it generates wheels for any dependencies that do not have wheels already uploaded for your system. So if numpy doesn't have a wheel for your arch, then it will build NumPy and put the wheel in the wheelhouse. If it's already on PyPI for your system, it will just use PyPI and will not add it to the built wheels in the wheelhouse. The wheelhouse + PyPI (or whatever index you are using) should be enough to install the current package without building anything.

**Comment by rth** [comment]
Thanks for the clarifications!

> This would be a mess

Yeah, I realize it's a complex subject, I'm clearly not suggesting it in this issue.

> That's because this is being misused, you should use build instead

Then https://packaging.python.org/guides/distributing-packages-using-setuptools/#wheels needs updating.

> But twine upload dist/* isn't that hard to type either,

It's not about ease of use but about confidence it won't upload any unrelated wheels (or versions) that happen to be in that folder. And also making it harder from a UI perspective not to upload a wheel it if it's already generated by `build`.

**Comment by rth** [comment]
> Then packaging.python.org/guides/distributing-packages-using-setuptools/#wheels needs updating.

Also unless I'm mistaken that page doesn't mention the `build` tool at all. I get that the path is "packaging with setuptools" but the title is very general "Packaging and distributing projects" and I think users would often end up there and not on the tutorial when searching.

**Comment by henryiii** [comment]
> Then ... needs updating

I agree. And there is too much duplication in packaging.python.org, IMO...

> So some logic that does the correct version filtering by default without relying on bash globing might help. As `poetry publish` does I imagine.

`poetry publish` reads pyproject.toml and pulls the name from there. Twine can't do that - what would you read? pyproject.toml PEP 621 metadata? setup.cfg? setup.py? You can't read a python file reliably, and running it can have side effects. Twine can upload files from other sources, too, not just setuptools. Due to the fact it is a general tool, I don't think you can make it much "smarter" than it currently is. Maybe a smart default based on PEP 621 would be interesting, though - something to discuss over on the twine issues.

**Comment by astrojuanlu** [comment]
> what would you read? pyproject.toml PEP 621 metadata? setup.cfg? setup.py? You can't read a python file reliably, and running it can have side effects.

My understanding is that https://github.com/pypa/pep517/ provides such unified interface. In fact, that's what pip-tools uses now to universally extract requirements from local packages, without the need of choosing which file to parse (have a look at https://github.com/jazzband/pip-tools/pull/1311, as well as this comment https://github.com/pypa/setuptools/issues/1951#issuecomment-722645440).

**Comment by henryiii** [comment]
Can someone open an issue in twine with the suggestion for the "smart" default, using `pep517.meta.load`, then?

**Comment by layday** [comment]
`pep517` provides an interface for extracting metadata from wheels. The question that remains is, how do you ascertain that the package name from the wheel is the same as in your `setup.cfg` or `pyproject.toml` or who knows what without rebuilding the wheel?

> It's not about ease of use but about confidence it won't upload any unrelated wheels (or versions) that happen to be in that folder. And also making it harder from a UI perspective not to upload a wheel it if it's already generated by `build`.

If `*`-globbing is too crude, you could search for distributions which begin with the name of your project: `twine publish dist/my-project-*.{tar.gz,whl}`.  See also https://github.com/pypa/build/issues/198 which attempts to address this in a different way.
